### PR TITLE
[Internal] First version of create flow with script + function in python

### PR DESCRIPTION
**Pull request by thy09**
# Description

Add a new `_script_executor` for execute a script with traces support.

Main changes:

* [`src/promptflow/promptflow/executor/_script_executor.py`](diffhunk://#diff-3294a0ccdf18c47385db91932bd860940cc3df5df10d206138fbcd48075530d5R1-R91): Added necessary imports to the `_script_executor.py` file.
* [`src/promptflow/promptflow/_core/flow_execution_context.py`](diffhunk://#diff-8a45b6238b72974b62aa211aec63ef4cbeadfa8277f84525442c245a16ee4461L41-R49): Modified the `__init__` method in the `FlowExecutionContext` class to have a default value for the `cache_manager` parameter.
* [`src/promptflow/promptflow/executor/flow_executor.py`](diffhunk://#diff-faa6c81d614b7e41b18a42a93139d961d92afa9aa9dd0b72cb6b7176d7541e69R177-R178): Modified the `create` method in the `FlowExecutor` class to add support for creating a `ScriptExecutor` instance based on the file extension. [[1]](diffhunk://#diff-faa6c81d614b7e41b18a42a93139d961d92afa9aa9dd0b72cb6b7176d7541e69R177-R178) [[2]](diffhunk://#diff-faa6c81d614b7e41b18a42a93139d961d92afa9aa9dd0b72cb6b7176d7541e69R163) [[3]](diffhunk://#diff-faa6c81d614b7e41b18a42a93139d961d92afa9aa9dd0b72cb6b7176d7541e69R190-R199)

Other changes:

* [`src/promptflow/promptflow/_core/run_tracker.py`](diffhunk://#diff-3ea41c85102f8318e1dc8568de0fb0de76ce92172e506df95e6f419808eabaf8R251-R252): Added a condition to the `end_run` method in the `RunTracker` class to only update the `api_calls` attribute if there are traces available.
* [`src/promptflow/tests/test_configs/eager_flows/dummy_flow_with_trace/flow.py`](diffhunk://#diff-0bf3e79590cc3678356ab2ff8087ceba5021c8e58cd399d8ce076072f83d8ef9R1-R24): Added a new file `flow.py` that defines functions for waiting, calling a dummy function, and creating tasks for different models.

# All Promptflow Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

**Review by D-W- — APPROVED** [review]

File: src/promptflow/promptflow/executor/_script_executor.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
+import asyncio
```
Comment (D-W-) [inline src/promptflow/promptflow/executor/_script_executor.py:1]: seems flake8 CI failed

File: src/promptflow/promptflow/executor/_script_executor.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        self._storage = storage or DefaultRunStorage()
+
+    def exec_line(
+        self,
+        inputs: Mapping[str, Any],
+        index: Optional[int] = None,
+        run_id: Optional[str] = None,
+        **kwargs,
+    ) -> LineResult:
+        operation_context = OperationContext.get_instance()
```
Comment (wangchao1230) [inline src/promptflow/promptflow/executor/_script_executor.py:57]: looks like we copied some code from flowExecutor, could we share more code? 

Ideally the execute dag part can be extract as a function in FlowExecutor, and for eager flow it's overrided as calling this part: 
image

File: src/promptflow/promptflow/executor/_script_executor.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        working_dir: Optional[Path] = None,
+        *,
+        storage: Optional[AbstractRunStorage] = None,
+    ):
+        logger.debug("Start initializing the executor.")
+        working_dir = Flow._resolve_working_dir(entry_file, working_dir)
+        m = load_python_module_from_file(entry_file)
+        self._func = getattr(m, str(func), None)
+        if not inspect.isfunction(self._func):
+            raise PythonLoadError(
```
Comment (D-W-) [inline src/promptflow/promptflow/executor/_script_executor.py:39]: Maybe add a test for this?

File: src/promptflow/promptflow/executor/_script_executor.py (hunk 2 above)
Comment (thy09) [reply src/promptflow/promptflow/executor/_script_executor.py:57]: There are a lot of stuffs we can improve, this is the first version to unblock other parts.

File: src/promptflow/promptflow/executor/_script_executor.py (hunk 3 above)
Comment (thy09) [reply src/promptflow/promptflow/executor/_script_executor.py:39]: Let's cover happy path first.

**Review by zzn2 — APPROVED** [review]

**Review by D-W- — APPROVED** [review]

**Review by wangchao1230 — APPROVED** [review]

File: src/promptflow/promptflow/executor/_script_executor.py (hunk 1 above)
Comment (thy09) [reply src/promptflow/promptflow/executor/_script_executor.py:1]: Fixed

**Pull request merged**